stack/provider: add new koding vars / add SSH key injecting during boostrap

rjeczalik opened the pull request:
This PR adds two following koding variables for use with stack templates:

- ${var.koding_stack_id}
- ${var.koding_template_id}

Supersedes #9542.

rjeczalik committed:
stack/provider: add var.koding_stack_id and var.koding_template_id

rjeczalik committed:
client: update google default template

rjeczalik committed:
client: update azure default template

ppknap commented on client/app/lib/providers/templates/google.coffee:
```diff
@@ -46,7 +46,7 @@ resource:
     google-instance:
       # and this is its identifier (required)
-      name: google-instance
+      name: 'koding-${var.koding_group_slug}-${var.koding_stack_id}-${count.index+1}'
```
Could you remove https://github.com/koding/koding/blob/c8c894efcb80d89cd4d49f43ef737a1bc148c289/go/src/koding/kites/kloud/provider/google/stack.go#L201 from google as well?

rjeczalik replied:
Done.

rjeczalik committed:
stack/provider: add SSHKeyPairFunc

This PR adds a SSHKeyPairFunc which can be used by
any provider to inject a newly generated SSH keypair
to their boostrap templates.

rjeczalik committed:
provider/google: drop appending count.index to name attr

rjeczalik committed:
provider/do: use unique SSH key-pair

rjeczalik commented:
I've added also a fix for Digital Ocean. Please do not merge it yet, as Digital Ocean most likely also suffers from the unique name problem, but I need to test it first to confirm.

rjeczalik committed:
stack: fix "invalid character '\n' in string literal" error

rjeczalik commented:
ptal

rjeczalik committed:
client: update digitalocean default template

rjeczalik commented:
ptal

ppknap commented:
@gokmen I wanted to test this change but I cannot initialize google provider 
![image](https://cloud.githubusercontent.com/assets/5021246/20350496/7696baf8-ac0f-11e6-87f7-2de719c0112e.png)

It works for aws, I assume this is something with credentials.

rjeczalik committed:
kloud/queue: remove excessive debug output

ppknap commented:
Credentials were added to mongo.
```
> db.jCredentials.find().pretty()                                                                                                                                            
{
        "meta" : {
                "createdAt" : ISODate("2016-11-16T13:56:13.854Z"),
                "modifiedAt" : ISODate("2016-11-16T13:56:13.854Z")
        },
        "verified" : true,
        "accessLevel" : "private",
        "provider" : "google",
        "title" : "Pawel google",
        "identifier" : "227c458b6969906418ad470d1ef4f411",
        "originId" : ObjectId("582c6546603100fc525e9d0e"),
        "_id" : ObjectId("582c657dadbc91165369221f")
}
```

rjeczalik commented:
> Credentials were added to mongo.

This is only jCredential, what about the actual credential?

ppknap commented:
@rjeczalik they are in social api and look correct:
![image](https://cloud.githubusercontent.com/assets/5021246/20352001/ca807112-ac15-11e6-8b6d-6571ee71ff6d.png)

rjeczalik committed:
credential: add debug printing

gokmen committed:
RequirementsParser: add new variable types

gokmen committed:
StackEditor: set the first usable provider as selected

gokmen committed:
ProvidersParser: ignore koding provider on credentials

rjeczalik commented:
@gokmen Thanks, works for me.

@cihangir @ppknap Ptal, so we can merge and deploy to sandbox.

ppknap commented on go/src/koding/tools/util/pretty.go:
```diff
@@ -0,0 +1,27 @@
+
+	p, err := json.MarshalIndent(v, "", "\t")
+	if err != nil {
+		return fmt.Sprintf("!JSON(%s)", []byte(p))
```
In case of `err != nil`, `p` will always be nil. Could you please not replace `p` during `MarshalIndent`?
